wagtail/query.py

```diff
@@ -515,37 +519,68 @@ def __iter__(self):
-                pk: ContentType.objects.get_for_id(pk) for _, pk in pks_and_types
-            }
+            # Convert above into a dict of lists. Used to query the database to find
+            # the correct pages for each type
+            pks_by_type_id = defaultdict(list)
+            for pk, type_id in pks_and_type_ids:
+                pks_by_type_id[type_id].append(pk)
 
-            # Get the specific instances of all pages, one model class at a time.
-            pages_by_type = {}
-            missing_pks = []
+            # A dict of ContentType instances keyed by id. Used to derive model
+            # classes from content type ids
+            ctypes_by_id = {
+                pk: ContentType.objects.get_for_id(pk) for pk in pks_by_type_id.keys()
+            }
 
-            for content_type, pks in pks_by_type.items():
+            proxy_ctype_ids = set()
+            for content_type in tuple(ctypes_by_id.values()):
+                model_class = content_type.model_class()
+                if model_class and model_class._meta.proxy:
+                    # Remember this is a proxy type
+                    proxy_ctype_ids.add(content_type.id)
+                    # Find and store the concrete 'parent' type
+                    parent_ctype = ContentType.objects.get_for_model(
+                        model_class, for_concrete_model=True
+                    )
+                    ctypes_by_id[parent_ctype.pk] = parent_ctype
+                    # Reshuffle pks_by_type_id so that proxy page instances with
+                    # shared parent models are fetched in the same query
+                    page_pks = pks_by_type_id.pop(content_type.pk, [])
+                    pks_by_type_id[parent_ctype.pk].extend(page_pks)
+
+            # Fetch specific pages, one concrete model class at a time, and store them
+            # in 'pages_by_type_id'
+            pages_by_type_id = defaultdict(dict)
+            missing_page_pks = set()
+
+            for ctype_id, pks in pks_by_type_id.items():
                 # look up model class for this content type, falling back on the original
                 # model (i.e. Page) if the more specific one is missing
-                model = content_types[content_type].model_class() or qs.model
+                model = ctypes_by_id[ctype_id].model_class() or qs.model
                 pages = model.objects.filter(pk__in=pks)
 
                 if qs._defer_streamfields:
                     pages = pages.defer_streamfields()
 
-                pages_for_type = {page.pk: page for page in pages}
-                pages_by_type[content_type] = pages_for_type
-                missing_pks.extend(pk for pk in pks if pk not in pages_for_type)
+                found_pks = set()
+                for page in pages:
+                    found_pks.add(page.pk)
+                    if page.content_type_id in proxy_ctype_ids:
+                        pages_by_type_id[page.cached_content_type.pk][
```

Comment:
question: isn't `page.cached_content_type.pk` the same as `page.content_type_id` https://github.com/wagtail/wagtail/blob/4c7e1ea1ce4c842ce30905c6539437fdf21be6df/wagtail/models/__init__.py#L1406 ?

Reply:
Good spot! That would certainly be simpler

Reply:
Resolved in 7ea93b8d13